File: ToolGood.Words/internals/TrieNode.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ToolGood.Words.internals
{
    public class TrieNode
    {
        //public byte Type { get; set; }
        public bool End { get; set; }
        public List<string> Results { get; private set; }
        internal Dictionary<char, TrieNode> m_values;
        private uint minflag = uint.MaxValue;
        private uint maxflag = uint.MinValue;


        public TrieNode()
        {
            m_values = new Dictionary<char, TrieNode>();
            Results = new List<string>();
        }

        public bool TryGetValue(char c, out TrieNode node)
        {
            if (minflag <= (uint)c && maxflag >= (uint)c) {
                return m_values.TryGetValue(c, out node);
            }
            node = null;
            return false;
        }
        public ICollection<TrieNode> Transitions()
        {
            return m_values.Values;
        }

        public TrieNode Add(char c)
        {
            TrieNode node;

            if (minflag > c) { minflag = c; }
            if (maxflag < c) { maxflag = c; }
            if (m_values.TryGetValue(c, out node)) {
                return node;
            }
            node = new TrieNode();
            m_values[c] = node;
            //m_values.Add(c, node);
            return node;
        }

        public void SetResults(string text)
        {
            if (End==false) {
                End = true;
            }
            Results.Add(text);
        }

        public void Merge(TrieNode node)
        {
            if (node.End) {
                if (End == false) {
                    End = true;
                }
                foreach (var item in node.Results) {
                    Results.Add(item);
                }
            }

            foreach (var item in node.m_values) {
                if (m_values.ContainsKey(item.Key) == false) {
                    if (minflag > item.Key) { minflag = item.Key; }
                    if (maxflag < item.Key) { maxflag = item.Key; }
                    m_values[item.Key] = item.Value;
                    //m_values.Add(item.Key, item.Value);
                }
            }
        }

    }

}

File: golang/TrieNode.go
package ToolGood_Words
 
type TrieNode struct {
	End bool  
	Results []string
	m_values map[int32]*TrieNode
	minflag int32 
	maxflag int32
}

func NewTrieNode() *TrieNode{
	return &TrieNode{
		End : false,
		m_values :  make(map[int32]*TrieNode) ,
		Results :   make([]string,0),
		minflag : 0,
		maxflag : 0xffff,
	} 
}

func (this *TrieNode) TryGetValue(c int32) (bool,*TrieNode){
	if this.minflag<=c && this.maxflag>=c {
		if val, s := this.m_values[c]; s {
			return true,val
		}
	}
	return false,nil
}

func (this *TrieNode) Add(c int32) *TrieNode{
	if val, s := this.m_values[c]; s {
		return val
	}
	if this.minflag<c {
		this.minflag = c
	}
	if this.maxflag>c {
		this.maxflag = c
	}
	node := NewTrieNode()
	this.m_values[c]=node
	return node 
}

func (this *TrieNode) SetResults(text string) {
	if this.End==false {
		this.End=true;
	}
	for i := 0; i < len(this.Results); i++ {
        if this.Results[i] == text {
            return
        }
    }
	this.Results=append(this.Results,text)
}

func (this *TrieNode) ToArray() *[]TrieNode{
	var first []TrieNode
	first =  make([]TrieNode,0xffff);
	for k,val:=range this.m_values{
		first[k] = *val
	}
	return &first
}
 

func (this *TrieNode) Merge(node *TrieNode,links map[*TrieNode]*TrieNode){
	if node.End==true{
		if this.End==false{
			this.End=true
		}
		for i := 0; i < len(node.Results); i++ {
			r:=node.Results[i]
			if StringsContains(this.Results,r)==-1 {
				this.Results=append(this.Results,r)
			}
		}
	}
	for key,val:=range node.m_values{
		if _, s2 := this.m_values[key]; s2 {
 		}else{
			if this.minflag<key {
				this.minflag = key
			}
			if this.maxflag>key {
				this.maxflag = key
			}
			this.m_values[key]=val
		}
	}
	if node2, s := links[node]; s {
		this.Merge(node2, links);
	}
}

func StringsContains(array []string, val string) (index int) {
    index = -1
    for i := 0; i < len(array); i++ {
        if array[i] == val {
            index = i
            return
        }
    }
    return
}
 